### ISSUE TITLE
Using condvar() for waiting on signal doesnt wake up

### ISSUE THREAD

**Issue by alekarasma** [OP-TEE/optee_os#3571]
Hello Everyone,

I am trying to use condvar in optee.

My code snippet is something like this:

```
REMV1 ttl0_command( const uint8_t* capdu, size_t capdu_size,
                         uint8_t* rapdu, size_t* rapdu_size )
{
  REMV1 rc;

  // Default bytes returned count to 0.
  *rapdu_size = 0;

  rc = ttl0_command_nb( capdu, capdu_size, 0 );
  if ( rc == REMV1_OK )
  {
    mutex_lock(&emv_ttl0_mutex);
    // Wait for completion.
    while ( !s_msg.finished )
      condvar_wait(&s_msg.cv, &emv_ttl0_mutex);

    mutex_unlock(&emv_ttl0_mutex);

    rc = s_msg.rc;
   }
  return rc;
}

static void request_process( imx_sim_rc_t rc, imx_sim_request_t* request )
{
  
  assert( !s_msg.finished );

  if ( rc != imx_sim_rc_ok )
  {
    pr_debug("SIM error %u\n", rc);
    s_msg.sim_rc = rc;
    s_msg.rc = emv1_rc_from_sim_rc( rc );
    s_msg.finished = true;
  }
  else
  {
    process( request );
    
    switch ( request->action )
    {
    case imx_sim_action_abort:
    case imx_sim_action_finalise:

      s_msg.finished = true;
      break;

    case imx_sim_action_tx:

      // Transmit & receive request.
      assert( request->tx.size != 0 );
      assert( request->rx_request != 0 );
      break;
    }
  }

  if ( s_msg.finished )
  {
    // Check that the return code was set.
    assert( s_msg.rc != REMV1_INVALID );

    condvar_signal(&s_msg.cv);
 }
 }
}

```
I am not signalling anything from interrupt but it still getting stuck. I have done mutex and condvar initialization in my init function. What can be the reason for this? Am I doing anything wrong?

Any help is really appreciated.

Thanks,
Asma

**Comment by jenswi-linaro** [comment]
You should hold the mutex `emv_ttl0_mutex` when calling `condvar_signal()`.

**Comment by alekarasma** [comment]
If I hold mutex while when calling condvar_signal(), 

```
 mutex_lock(&emv_ttl0_mutex);
  if ( s_msg.finished )
  {
     
    condvar_signal(&emv_ttl0_cv);
    mutex_unlock(&emv_ttl0_mutex);
    // Check that the return code was set.
    assert( s_msg.rc != REMV1_INVALID );

    if ( s_msg.rc != REMV1_OK )
    {
      s_session.initialised = false;
    }
```

 I get the following output :

> E/TC:0   assertion 'thread_get_id_may_fail() != -1' failed at core/arch/arm/kernel/mutex.c:22 <__mutex_lock>
> E/TC:0   Panic at core/kernel/assert.c:28 <_assert_break>
> E/TC:0   Call stack:
> E/TC:0    0xbe00577d
> 

And I think the reason behind this is, The above function is callback function. Is my understanding correct?

Regards,
Asma

**Comment by jenswi-linaro** [comment]
> I get the following output :
> 
> > E/TC:0   assertion 'thread_get_id_may_fail() != -1' failed at core/arch/arm/kernel/mutex.c:22 <__mutex_lock>
> > E/TC:0   Panic at core/kernel/assert.c:28 <_assert_break>
> > E/TC:0   Call stack:
> > E/TC:0    0xbe00577d
> 
> And I think the reason behind this is, The above function is callback function. Is my understanding correct?

These functions are only valid to call from a normal thread. Interrupt context or from a fast call is not OK. The reason for this is that an RPC may be needed to wake up an eventual waiter, and RPC is only possible from a normal thread context.

**Comment by alekarasma** [comment]
Thanks @jenswi-linaro

**Comment by angelina-zhao** [comment]
> > I get the following output :
> > > E/TC:0   assertion 'thread_get_id_may_fail() != -1' failed at core/arch/arm/kernel/mutex.c:22 <__mutex_lock>
> > > E/TC:0   Panic at core/kernel/assert.c:28 <_assert_break>
> > > E/TC:0   Call stack:
> > > E/TC:0    0xbe00577d
> > 
> > 
> > And I think the reason behind this is, The above function is callback function. Is my understanding correct?
> 
> These functions are only valid to call from a normal thread. Interrupt context or from a fast call is not OK. The reason for this is that an RPC may be needed to wake up an eventual waiter, and RPC is only possible from a normal thread context.

I met the similar problem recently.  There is a HW AES engine in my SoC, which supports interrupt.  I expected to use the interrupt mode to avoid polling in my design, such as:

1, in normal thread, call aes_start then wait
2, in the HW interrupt, signal and wake up the normal thread

However, I notice condvar uses mutex, and mutex uses thread_rpc(), which cannot be used in interrupt mode.

Is polling the only way for secure hardware in current OPTEE OS design?  And if so, do you have plan to support completion/wait for interrupt in the future?

Thanks.

@jenswi-linaro

**Comment by jenswi-linaro** [comment]
> Is polling the only way for secure hardware in current OPTEE OS design?

Yes, unfortunately that's still the case.

> And if so, do you have plan to support completion/wait for interrupt in the future?

We want support interrupt driven drivers in some way. The current assumption is that it could be done with split drivers where a top-half receives the interrupt and then notifies a sleeping thread that there's work to be done. The crux is how to notify a thread sleeping in normal world. One option is to use and SGI which would be handled as an non-secure interrupt by normal world. Once in normal world it wouldn't be hard to wake up a sleeping thread. However, the Linux kernel is already using its SGIs for other purposes. One way or the other we need to change Linux kernel SGI usage, in the end we'll probably reallocate one secure SGI to be used by normal world instead.

Currently we don't have any driver in upstream OP-TEE that need this. Until we do it will probably we quite hard to introduce the needed changes in the Linux kernel since we don't have a solid use-case. What makes matters worse is that we'd likely need to separate solutions in the Linux kernel, one for AArch32 and one for AArch64. We may also need to take bootloaders (U-Boot and EDK2) into account, but perhaps not in the first step.

**Comment by angelina-zhao** [comment]
Thank you for the quick reply.

I will figure out whether I can to some private changes on my system (may be some SGI are never used in my real case) to implement the interrupt driven driver, as I do not need to take care of the common solution for all platforms. : )

**Comment by jenswi-linaro** [comment]
Good luck! I hope you'll be able to sort something out. If you do, would you mind sharing a bit of the solution? It doesn't have to be patches.

**Comment by angelina-zhao** [comment]
sure, I will come back when I have any achievements on this.

**Comment by angelina-zhao** [comment]
> Good luck! I hope you'll be able to sort something out. If you do, would you mind sharing a bit of the solution? It doesn't have to be patches.

Your suggestion works.

I use a spare SPI interrupt in GIC, and trigger an REE interrupt when HW AES Engine TEE interrupt arrives, such as:

static enum itr_return seceng_itr_handler(struct itr_handler *handler)
{
	(void)handler;

	...
	//Clear the HW interrupt.

	//update the flag to bypass the wait in case wait() is called after interrupt is triggered.
	old_itr_status = cpu_spin_lock_xsave(&wq_spin_lock);
	crypto_wait_status = 1;
	cpu_spin_unlock_xrestore(&wq_spin_lock, old_itr_status);
	
	//raise irq to REE.
	itr_raise_pi(IRQ_SECENG_CRYPTO_WQ_ID);

	return ITRR_HANDLED;
}
KEEP_PAGER(seceng_itr_handler);

And the wait() rountine is as below.  Before the thread exception is ready, polling is still used.  After that, the wait will return to REE with a hard-coded cookie 0x12345678;

static int crypto_aes_wait(void)
{
	uint32_t val;
	uint32_t old_itr_status;

	val = thread_get_exceptions();
	if ((val&THREAD_EXCP_NATIVE_INTR) == 0) {
		if (crypto_wait_status != 0) {
			__wq_rpc(0, 0x12345678, NULL, 0, 0);
		} else {
			DMSG("crypto_aes_wait wq already done...exit.");
		}
	} else {
		DMSG("crypto_aes_wait interrupt not enabled.");
		... // still polling here.
	}

	old_itr_status = cpu_spin_lock_xsave(&wq_spin_lock);
	crypto_wait_status = 0;
	cpu_spin_unlock_xrestore(&wq_spin_lock, old_itr_status);

	return 0;
}

In Linux , I add the interrupt registration in tee_core.c, and bind a handler to tee-supp device.

irqreturn_t tee_wq_interrupt(int irq, void *dev_cookie)
{
	struct optee * optee = (struct optee *)dev_cookie;
	wq_wakeup(&optee->wait_queue, 0x12345678);
	return IRQ_HANDLED;
}

**Comment by jenswi-linaro** [comment]
@angelina-zhao, thanks for sharing
